server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -342,29 +342,36 @@ protected void createMissingIndicesAndIndexData(
         final AtomicArray<BulkItemResponse> responses = new AtomicArray<>(bulkRequest.requests.size());
         // Optimizing when there are no prerequisite actions
         if (indicesToAutoCreate.isEmpty() && dataStreamsToBeRolledOver.isEmpty() && failureStoresToBeRolledOver.isEmpty()) {
-            executeBulk(task, bulkRequest, startTimeNanos, listener, executor, responses, Map.of());
+            executeBulk(task, bulkRequest, startTimeNanos, listener, executor, responses);
             return;
         }
-        final Map<String, IndexNotFoundException> indicesThatCannotBeCreated = new HashMap<>();
+        Map<String, Exception> indicesExceptions = new ConcurrentHashMap<>();
+        Map<String, Exception> dataStreamExceptions = new ConcurrentHashMap<>();
+        Map<String, Exception> failureStoreExceptions = new ConcurrentHashMap<>();
```

Comment:
To avoid the scenario where there are two requests in a `BulkRequest` where one goes to a data stream that is marked for rollover and another goes to that data stream's failure store that is marked for rollover, and both rollovers fail, we have two separate maps. If we'd have one map, one of those failed rollovers would overwrite the other in the map (because the key (i.e. data stream name) would be the same).